### PR TITLE
Adding an example simulation test.

### PR DESCRIPTION
**Pull request by palafranchise**
*Description of changes:*

This change adds a test_nodes package with an example test node for evaluating robot behavior in simulation. The test node asserts that the robot can navigate to its goal successfully. The test node uses the Tag and Cancel hosted-ros-node services to record test status and self-terminate. New launch files are added for running the robot code in the small house world. The number of successful navigation to check for and the simulation world to run in can be configured with environment variables. If the environment variables are not set the defaults are '1' and 'bookstore'. The launch file and code exercised by the "1-click" workflow are left unchanged.

These changes have been tested by invoking the test_world launch file in AWS RoboMaker simulation to trigger 4 tests: 
- a single navigation in the bookstore world
- 4 navigations in the bookstore world
- a single navigation in the small house world 
- 4 navigations in the small house world

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: simulation_ws/src/test_nodes/nodes/navigation_test [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    def is_complete(self):
+        return self.successful_navigations >= self.navigation_success_count
+
+    def check_complete(self, msgs):
+        for msg in msgs.metrics:
+            if msg.metric_name == METRIC_NAME:
+                rospy.loginfo("Metric Name: %s", msg.metric_name)
+                rospy.loginfo("Metric Value: %s", msg.value)
+                if msg.value == 0.0 and self.is_completed == False and self.latch == False:
+                    self.set_latched()
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/nodes/navigation_test]: why do we latch?

File: simulation_ws/src/test_nodes/nodes/navigation_test [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    
+    def is_complete(self):
+        return self.successful_navigations >= self.navigation_success_count
+
+    def check_complete(self, msgs):
+        for msg in msgs.metrics:
+            if msg.metric_name == METRIC_NAME:
+                rospy.loginfo("Metric Name: %s", msg.metric_name)
+                rospy.loginfo("Metric Value: %s", msg.value)
+                if msg.value == 0.0 and self.is_completed == False and self.latch == False:
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/nodes/navigation_test]: You'll want some tolerance here -- distance to goal is a floating point _and_ depends on the robot nav stack. 

Something like:  `abs(msg.value) < tolerance` where `tolerance` is perhaps 25cm or so

File: simulation_ws/src/test_nodes/nodes/navigation_test [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    self.set_latched()
+                    self.increment_navigations()
+                    self.utils.set_tag(name=self.test_name + "_Successful_Nav_" + str(self.successful_navigations), value=str(self.successful_navigations))
+                    if self.is_complete():
+                        self.is_completed = True
+                        self.utils.set_tag(name=self.test_name, value="Passed")
+                        self.utils.cancel_job()
+                elif msg.value > 1 and self.is_completed == False:
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/nodes/navigation_test]: Bit confused why you need this check -- related to latching.

File: simulation_ws/src/test_nodes/nodes/navigation_test [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import unittest
+
+from rosgraph_msgs.msg import Clock
+from ros_monitoring_msgs.msg import MetricList
+from robomaker_simulation_msgs.msg import Tag
+from robomaker_simulation_msgs.srv import Cancel, AddTags
+
+METRIC_NAME = "distance_to_goal"
+
+class SimulationUtils():
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/nodes/navigation_test]: Not sure why you need a class. The state is never used.  Free functions seem sufficient.

File: simulation_ws/src/test_nodes/nodes/navigation_test [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                        self.utils.cancel_job()
+                elif msg.value > 1 and self.is_completed == False:
+                    self.set_unlatched()
+                    
+    def test_navigation(self):
+    	try:
+    		self.utils.set_tag(name=self.test_name + "_Time_Elapsed_Start", value= str(time.time()).split(".", 1)[0])
+    		self.clock = rospy.Subscriber('/metrics', MetricList, self.check_complete)
+    		rospy.spin()
+    	except:
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/nodes/navigation_test]: might need a comment: does it fall out here always unless the `is_complete()` is trigged? I.e, when the job expires the `except:` is invoked?

Comment (samuelgundry) [reply simulation_ws/src/test_nodes/nodes/navigation_test]: Separately, log the exceptions here to help debugging.

File: simulation_ws/src/test_nodes/nodes/navigation_test [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            self.utils.set_tag(name=self.test_name + "_Time_Elapsed_Status", value="Failed")
+            self.utils.cancel_job()
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/nodes/navigation_test]: After you cancel the job, if this node exists and is required, then the Job will have an APP_CRASH failure. The recommended practice  is to cancel and then sleep. However, you don't want to do that in a crash situation. I would cancel and sleep on expected non-crash situations like a timeout and then on all others, log the exception, set the tag, and just exit.  Thoughts?

**Review by samuelgundry — CHANGES_REQUESTED** [review]

File: simulation_ws/src/test_nodes/nodes/navigation_test (hunk 4 above)
Comment (palafranchise) [reply simulation_ws/src/test_nodes/nodes/navigation_test]: Fair, removed.

File: simulation_ws/src/test_nodes/nodes/navigation_test (hunk 6 above)
Comment (palafranchise) [reply simulation_ws/src/test_nodes/nodes/navigation_test]: Cancel job here is sending a message to the hosted-ros-node, we don't actually exit. I think this is the pattern we want to recommend--record the state of the test and let the API shut the sim down. A comment would be good, will add.

File: simulation_ws/src/test_nodes/nodes/navigation_test (hunk 5 above)
Comment (palafranchise) [reply simulation_ws/src/test_nodes/nodes/navigation_test]: Will add a comment. This except: is fine (will add error logging) but we do need a base timeout to catch cases were the robot gets stuck. Will add a wall clock timeout and make it configurable with a default.

File: simulation_ws/src/test_nodes/nodes/navigation_test (hunk 3 above)
Comment (palafranchise) [reply simulation_ws/src/test_nodes/nodes/navigation_test]: Synced offline, latching is overloaded here. This is not latching as in ros topic "Latching mode". Sorry for the confusion. We're creating a latch to catch the case when the distance_to_goal metric produces multiple values (which is always) so we can properly emit a single tag. Will add a comment. And will add a comment on the merits of using this metric verses a ground truth value. I'll also check for a non-zero range so the example reflects what you'd likely need to with real-world data/nav stack.

File: simulation_ws/src/test_nodes/nodes/navigation_test (hunk 2 above)
Comment (palafranchise) [reply simulation_ws/src/test_nodes/nodes/navigation_test]: This was my thought initially as well but once I took a look at distance_to_goal I realized 0.0 is always and repeatedly published as the goal is reached. I'll check for a non-zero range so the example reflects what you'd likely need to with real-world data/nav stack.

File: simulation_ws/src/test_nodes/nodes/navigation_test (hunk 1 above)
Comment (palafranchise) [reply simulation_ws/src/test_nodes/nodes/navigation_test]: We're subscribed to /metrics and distance_to_goal publishes 0 multiple times when the robot reaches/nears the goal. Latching ensures that we only publish the successful nav and overall success once.

File: simulation_ws/src/test_nodes/nodes/navigation_test (hunk 4 above)
Comment (jerwallace) [reply simulation_ws/src/test_nodes/nodes/navigation_test]: I think in a perfect world, we would have a "simulation utils" library that customers could include with their applications, which would likely store some context (and state) about the current simulation job... Interested in your thoughts on this! My original thought for using a class was to create extensible code (so we could easily add more operations in the future) where customers could easily pull the utilities out and use them in their applications.

Perhaps, we should just break out the utilities class as an external python file for now and reference it as an include in the NavigationTest? These two operations (cancel and tag) are not related to the test itself, so makes sense for them to be separate.

File: simulation_ws/src/test_nodes/launch/test.launch [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+       See http://wiki.ros.org/Clock
+
+       Note: set to false for deploying to a real robot.
+  -->
+  <arg name="use_sim_time" default="true"/>
+  <param name="use_sim_time" value="$(arg use_sim_time)"/>
+
+  <!-- Rotate the robot on launch -->
+  <param name="CLOCK_TIMEOUT_SECONDS" type="int" value="$(optenv CLOCK_TIMEOUT_SECONDS 300)" />
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/launch/test.launch]: nit; make sure it's clear this is Sim time, not wallclock

File: simulation_ws/src/test_nodes/nodes/navigation_test [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    
+    def is_complete(self):
+        return self.successful_navigations >= self.navigation_success_count
+        
+    def check_timeout(self, msg):
+        """
+            Cancel the test if it times out. The timeout is based on the
+            /clock topic (simulation time).
+        """
+        if msg.clock.secs > self.timout and not self.is_cancelled:
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/nodes/navigation_test]: typo "timeout" vs "timout"

File: simulation_ws/src/test_nodes/nodes/navigation_test [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                """
+                    If our distance to goal metric drops below 1 and we've
+                    achieved our goal count then we are complete and we tag the
+                    job success and cancel. Else, we continue checking progress
+                    towards a new goal once the distance to goal climbs back 
+                    above 1. Note that we're using what the nav stack thinks
+                    is the distance to goal, in the real world we'd want to use
+                    a ground truth value to ensure accuracy. 
+                """
+                if msg.value <= 1 and self.is_completed == False and self.latch == False:
```
Comment (samuelgundry) [inline simulation_ws/src/test_nodes/nodes/navigation_test]: be aware the value could fluctuate around 1. Should use a range (which you had before but assumed it triggered at == 0.0)

**Review by samuelgundry — COMMENTED** [review]
Looks good. Typo withstanding.  I misunderstand the latching/range in the first revision, and conflated the == 0.0 vs. 0 < value < 1.0 usage. See comments.

**Review by raghaprasad — APPROVED** [review]
LGTM

- New line
- Fix file name
- Follow up with fixing defaults in launch file.

**Review by samuelgundry — APPROVED** [review]
Nice. Thanks for the revisions.

**Pull request merged**